### ISSUE TITLE
401 Unauthorized after recreation of Rdio object

### ISSUE THREAD

**Issue by anbutle** [rdio/api#130]
This is in reference to the Android API.

Came across this while writing some unit tests. It seems that if I recreate an Rdio object shortly after disposing of one (within about 1 min), my api requests on the new object all return 401 Unauthorized. In my scenario I setup one unit test and create an Rdio object, make a call to play my collection station for 30 seconds, and then tear down the test. The next test starts, creates a new Rdio object (which claims "Rdio Api initialized") but then my next calls "getCurrentUser and getPlaylists" fail with a 401. I am using the same token and secret both times.

Ideally I would like to be able to create a new Rdio object for each test to check some stuff from a clean state, so this is causing some trouble. Perhaps there is something I am missing?

Thanks

**Comment by anbutle** [comment]
I wanted to update as I got a chance to do a little more research into this tonight. I traced the api calls that were being caused by my tests. Here is a breakdown of the steps that occured.
1 - Setup unit test one (creates Rdio object using my credentials, appkey and secret).
2 - Run unit test one (test passes)
3 - Tear down unit test one (kill some helper objects and call cleanup on Rdio object).
4 - Set up unit test two (call same method as case 1 to create Rdio object with credentials). During the setup the following 3 calls are made.
5 - http trace shows that this caused a call to getPlaybackToken, which succeeded
6 - next call was to getPlaylists, which succeeded
7 - next call was to getUser, which failed

The failure in step 7 seems to be due to my oauth data using the same nonce value as the call in 6. The 2 requests are identical (except for content-length, and the actual data).

In my app, the first thing I do when I get onRdioReady() called is call getUser and getPlaylists, I make these calls at the same time, one after the other. It seems that sometimes this causes issues with oauth and I get the "401 Duplicate Nonce" response. I have changed my app so that the getPlaylists call occurs once getUser returns which seems to have corrected my problem. 

It seems to me like there is an issue with making concurrent calls, or at least calls executed one after the other. I was using charles proxy to capture api request traces and can provide them if it helps. 

Thanks

**Comment by anbutle** [comment]
Wanted to follow up. I was getting 401s all over the place in my application so I decided to create a proper fix to avoid the scenarios that triggers the bug. I wrote a simple dispatch queue to control the execution of API calls to avoid duplicate nonce and invalid signature errors with OAuth. Using the RdioApiDispatchQueue gist below I no longer got 401s. 

https://gist.github.com/rebutl/72a1f6f19636960754fe
